lib/pages/trade/bloc/TradePageBloc.dart

```diff
@@ -0,0 +1,189 @@
+    on<SetTokenTo>(_mapSetTokenToEventToState);
+    on<SwapTokens>(_mapSwapTokensEventToState);
+  }
+
+  get initialState => TradePageState();
+
+  void _mapRefreshEventToState(
+      PageRefreshEvent event, Emitter<TradePageState> emit) async {
+    emit(state.copyWith(status: Status.loading));
+    try {
+      final tokenFromBalance = await walletController
+          .getTokenBalance(state.tokenFrom!.address.value);
+      final tokenToBalance =
+          await walletController.getTokenBalance(state.tokenTo!.address.value);
+      emit(state.copyWith(
+          tokenFromBalance: double.parse(tokenFromBalance),
+          tokenToBalance: double.parse(tokenToBalance)));
+      final response = await repo.fetchSwapInfo(
+          tokenFrom: state.tokenFrom!.address.value,
+          tokenTo: state.tokenTo!.address.value);
+      final isSuccess = response.isLeft();
+      print("isSuccess - $isSuccess");
+      if (isSuccess) {
+        swapController.updateFromAddress(state.tokenFrom!.address.value);
+        swapController.updateToAddress(state.tokenTo!.address.value);
+        final swapInfo = response.getLeft().toNullable()!.swapInfo;
+        final transactionInfo =
+            calculateTransactionInfo(swapInfo, state.tokenInputFromAmount);
+
+        //do some math
+        emit(state.copyWith(
+          status: Status.success,
+          minimumReceived: transactionInfo.minimumReceived!.toDouble(),
+          priceImpact: transactionInfo.priceImpact!.toDouble(),
+          receiveAmount: transactionInfo.receiveAmount!.toDouble(),
+          price: transactionInfo.price!.toDouble(),
+          lpFee: transactionInfo.lpFee!.toDouble(),
+        ));
+      } else {
+        final errorMsg = response.getRight().toNullable()!.errorMsg;
+        //TODO Create User facing error messages https://athletex.atlassian.net/browse/AX-466
+        print(errorMsg);
+        emit(state.copyWith(
+          status: Status.error,
+          minimumReceived: 0,
+          priceImpact: 0,
+          receiveAmount: 0,
+          price: 0,
+          lpFee: 0,
+        ));
+      }
+    } catch (e) {
+      emit(state.copyWith(status: Status.error));
+    }
+  }
+
+  SwapTransactionInfo calculateTransactionInfo(
+      SwapInfo? swapInfo, double inputAmount) {
+    final toReserve = swapInfo!.toReserve;
+    final fromReserve = swapInfo.fromReserve;
+    final fromInputAmount = inputAmount;
+    final receiveAmount =
+        fromInputAmount * (toReserve / (fromReserve + fromInputAmount));
+    final price = swapInfo.toPrice;
+    final lpFee = fromInputAmount * 0.003;
+    final priceImpact = 100 *
+        (1 -
+            ((fromReserve * (toReserve - receiveAmount)) /
+                (toReserve * (fromReserve + fromInputAmount))));
+
+    final minimumReceiveAmt = receiveAmount * (1 - _slippage_tolerance);
+    return SwapTransactionInfo(
+        minimumReceiveAmt, priceImpact, receiveAmount, price, lpFee);
+  }
```

Comment:
Is this the same logic as the BuyDialogBlocs function? If so we need to extract it to a dependency so we can do it once and for all for all and have one place that we test it.

Reply:
Buy and Sell is a case of Swap so it uses the same calculations.